docket/core/sync_async.py

```diff
@@ -12,17 +12,28 @@ def __init__(self, manager: SyncAsync, awaitable: Awaitable[_T]) -> None:
         self.awaitable = awaitable
         self.manager = manager
         self.responded: bool = False
+        self.errored: bool = False
         self.response: _T
+        self.exc: Exception
 
     async def complete(self) -> None:
-        self.set_result(await self.awaitable)
+        try:
+            ret = await self.awaitable
+        except Exception as e:
+            self.set_error(e)
+        else:
+            self.set_result(ret)
 
     def set_result(self, result: _T) -> None:
         self.response = result
         self.responded = True
 
+    def set_error(self, exc: Exception) -> None:
+        self.exc = exc
+        self.errored = True
```

Comment:
I dont see a reason to use two variables. It can just be an optional one. Same goes for responded and response.

Reply:
I'm not sure what you mean - could you maybe show an example?

Reply:
```python
# In __init__
response: Optional[_T] = None
exc: Optional[Exception] = None

# In async wait
while not (response or exc):
    sleep(0.1)
```

Reply:
What if the callback returns None?

Reply:
Actually I have a solution for this.